sympy/core/mul.py

```diff
@@ -1353,20 +1358,21 @@ def _eval_is_even(self):
             return False
 
     def _eval_is_composite(self):
-        if self.is_integer and self.is_positive:
```

Comment:
This check is redundant because we check that each arg is an integer below. It might seem like the check above is a shortcut but it also loops through all args checking that they are integers.

In general it is important that an _eval_is_y method doesn't use self.is_x (it should check properties on args not self) although there are exceptions. It becomes problematic when self.is_x comes back to self.is_y because we have what would be a recursion error except for this line:
https://github.com/sympy/sympy/blob/master/sympy/core/assumptions.py#L283
Although that line stops recursion errors it leads to other weirdness like getting a different value for the same property when accessing it multiple times. The fix is to ensure that _eval_is methods only query properties of args rather than self.